lib/test-result-formatting.js

```diff
@@ -0,0 +1,95 @@
+'use strict';
+
+var chalk = require('chalk');
+var _ = require('underscore'); // TODO: `lodash` is more performant and more widely used these days
+
+var formatSuccessLine = function(success) {
+  return success['name'] + '#' + success['methodName'] + ' took ' + success['time'];
+};
+
+var formatSuccesses = function(testResult) {
+  var successes = testResult['successes'];
+
+  if (!successes || !successes.length) { return []; }
+
+  return ['Successes:'].concat(_.map(
+    successes, _.compose(chalk.green, formatSuccessLine)
+  ));
+};
+
+var formatFailure = function(failure) {
+  return [
+    chalk.red(failure['name'] + '#' + failure['methodName'] + ' took ' + failure['time']),
+    '  - ' + failure['message'],
+    '  - ' + failure['stackTrace']
+  ];
+};
+
+var formatFailures = function(testResult) {
+  var failures = testResult['failures'];
+
+  if (!failures || !failures.length) { return []; }
+
+  return _.flatten(['Failures:'].concat(_.map(failures, function (failure) {
+    return formatFailure(failure);
+  })));
+};
+
+var formatOverview = function(testResult) {
+  var successCount = (testResult['successes'] && testResult['successes'].length) || 0;
+  var failureCount = (testResult['failures'] && testResult['failures'].length) || 0;
+  var testCount = successCount + failureCount;
+  var testCountText = testCount + ' method' + (testCount > 1 ? 's' : '');
+  var failureCountText = failureCount + ' failures';
+
+  return [testCountText + ', ' + failureCountText];
+};
+
+var formatCodeCoverageWarnings = function(testResult) {
+  var warnings = testResult.codeCoverageWarnings;
+
+  // TODO: What is this for? Or "When is this true?"
```

Comment:
Yes, this is required! :-/
Unfortunately (all) the Metadata API responses return an object instead of an array if there is only one item. For the Tooling API this doesn't seem to be the case.

See https://gist.github.com/amtrack/d978987b1789e1f97c2285f6d29ecace for some sample responses for running `force-dev-tool test`.

So we should have a function s.th. like this

```javascript
var ensureArray = function(objectOrArray) {
  if (!objectOrArray) {
    return [];
  } else if (!Array.isArray(objectOrArray) {
    return [objectOrArray];
  }
  return objectOrArray;
};

var xxx = ensureArray(testResult.xxx);
if (!xxx.length) { return []; }
...
// return array with heading and lines
```